lib/bugsnag/session_tracker.rb

```diff
@@ -41,117 +36,93 @@ def create_session
         }
       }
       SessionTracker.set_current_session(new_session)
-      add_thread = Thread.new { add_session(start_time) }
-      add_thread.join()
+      add_session(start_time)
     end
 
-    def send_sessions
-      @mutex.lock
-      begin
-        deliver_sessions
-      ensure
-        @mutex.unlock
-      end
-    end
-
-    private
-    def add_session(min)
-      @mutex.lock
-      begin
-        @registered_at_exit = false unless defined?(@registered_at_exit)
-        if !@registered_at_exit
-          @registered_at_exit = true
-          at_exit do
-            if !@deliver_fallback.nil? && @deliver_fallback.status == 'sleep'
-              @deliver_fallback.terminate
-            end
-            deliver_sessions
-          end
-        end
-        @session_counts[min] ||= 0
-        @session_counts[min] += 1
-        if Time.now() - @last_sent > TIME_THRESHOLD
-          deliver_sessions
-        end
-      ensure
-        @mutex.unlock
-      end
-    end
+    alias :create_session :start_session
 
-    def deliver_sessions
-      return unless @config.track_sessions
+    def send_sessions
       sessions = []
-      @session_counts.each do |min, count|
+      counts = @session_counts.dup
```

Comment:
You actually want

```ruby
counts = @session_counts
@session_counts = Concurrent::Hash.new(0)
counts.each ...
```

With the current implementation there is still a race condition.